### PR TITLE
Move from LibGit2Sharp v2.4.0 to v2.3.1

### LINKED ISSUE #1244: Errors opening a pull request: "Failed to nmap..."
Environment:
GitHub Enterprise
Git repo has two remotes. One is a fork.
VS 2017 15.3.5
GitHub.VisualStudio 2.3.3.38

Repro:
1) Open GitHub repo
2) Click Pull Requests
3) Select the fork remote
4) Select any pull request

Expected:
Opens PR

Actual:
Briefly spins and then displays the following error:

'Failed to nmap. No data written: Not enough storage is available to process this command.'

### PR DESCRIPTION
**Pull request by jcansdale**
We upgraded from LibGit2Sharp v0.22 to v0.24 in PR #1248. This upgrade was in order to fix #1244 (the `Failed to mmap` issue).

Unfortunately, LibGit2Sharp v2.4.0 sometimes throws an `AccessViolationException` when `RetrieveStatus` is called (see #1315). This appears to still be an issue in v0.25.0-preview-0033.

Moving to LibGit2Sharp v2.3.1 because it appears to be more stable than v0.24 and fixes both #1244 and #1315.

NOTE: I've left in some of the `dependentAssembly` and `NuGetPackageImportStamp` junk that NuGet likes to add. Hopefully this will make the next time we update NuGet packages cleaner. We don't package the `App.config` files, so this shouldn't make any difference to the extension.

Workaround for: #1306

### CODE REVIEW

File: src/DesignTimeStyleHelper/App.config [hunk 1]
```diff
@@ -45,6 +45,14 @@
...
         <bindingRedirect oldVersion="0.0.0.0-4.0.1.0" newVersion="4.0.1.0" />
       </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="Microsoft.VisualStudio.ComponentModelHost" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-14.0.0.0" newVersion="14.0.0.0" />
+      </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="Microsoft.VisualStudio.Validation" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-14.0.0.0" newVersion="14.0.0.0" />
+      </dependentAssembly>
```
Comment (shana) [inline src/DesignTimeStyleHelper/App.config]: These changes shouldn't be in this PR

File: src/GitHub.App/GitHub.App.csproj [hunk 2]
```diff
@@ -16,6 +16,8 @@
     <CodeAnalysisRuleSet>..\common\GitHubVS.ruleset</CodeAnalysisRuleSet>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <CodeAnalysisIgnoreGeneratedCode>true</CodeAnalysisIgnoreGeneratedCode>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
```
Comment (shana) [inline src/GitHub.App/GitHub.App.csproj]: These changes shouldn't be here

File: src/GitHub.Exports.Reactive/GitHub.Exports.Reactive.csproj [hunk 3]
```diff
@@ -15,6 +15,8 @@
     <CodeAnalysisRuleSet>..\common\GitHubVS.ruleset</CodeAnalysisRuleSet>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <CodeAnalysisIgnoreGeneratedCode>true</CodeAnalysisIgnoreGeneratedCode>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
```
Comment (shana) [inline src/GitHub.Exports.Reactive/GitHub.Exports.Reactive.csproj]: Same as above

File: src/GitHub.Exports.Reactive/GitHub.Exports.Reactive.csproj [hunk 4]
```diff
@@ -171,7 +173,9 @@
     <Compile Include="Validation\ReactivePropertyValidator.cs" />
   </ItemGroup>
   <ItemGroup>
-    <None Include="packages.config" />
+    <None Include="packages.config">
+      <SubType>Designer</SubType>
+    </None>
```
Comment (shana) [inline src/GitHub.Exports.Reactive/GitHub.Exports.Reactive.csproj]: Same as above

File: src/GitHub.Exports/GitHub.Exports.csproj [hunk 5]
```diff
@@ -15,6 +15,8 @@
     <CodeAnalysisRuleSet>..\common\GitHubVS.ruleset</CodeAnalysisRuleSet>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <CodeAnalysisIgnoreGeneratedCode>true</CodeAnalysisIgnoreGeneratedCode>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
```
Comment (shana) [inline src/GitHub.Exports/GitHub.Exports.csproj]: Same as above

File: src/GitHub.Exports/GitHub.Exports.csproj [hunk 6]
```diff
@@ -210,7 +212,9 @@
     <Compile Include="..\common\SolutionInfo.cs">
       <Link>Properties\SolutionInfo.cs</Link>
     </Compile>
-    <None Include="packages.config" />
+    <None Include="packages.config">
+      <SubType>Designer</SubType>
+    </None>
```
Comment (shana) [inline src/GitHub.Exports/GitHub.Exports.csproj]: Same as above

File: src/GitHub.StartPage/app.config [hunk 7]
```diff
@@ -26,6 +26,10 @@
         <assemblyIdentity name="Microsoft.VisualStudio.CoreUtility" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
         <bindingRedirect oldVersion="0.0.0.0-15.0.0.0" newVersion="15.0.0.0" />
       </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="Microsoft.VisualStudio.ComponentModelHost" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-14.0.0.0" newVersion="14.0.0.0" />
+      </dependentAssembly>
```
Comment (shana) [inline src/GitHub.StartPage/app.config]: Same as above

File: src/GitHub.TeamFoundation.14/GitHub.TeamFoundation.14.csproj [hunk 8]
```diff
@@ -17,6 +17,8 @@
     <CodeAnalysisRuleSet>..\common\GitHubVS.ruleset</CodeAnalysisRuleSet>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <CodeAnalysisIgnoreGeneratedCode>true</CodeAnalysisIgnoreGeneratedCode>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
```
Comment (shana) [inline src/GitHub.TeamFoundation.14/GitHub.TeamFoundation.14.csproj]: Same as above

File: src/GitHub.TeamFoundation.15/GitHub.TeamFoundation.15.csproj [hunk 9]
```diff
@@ -17,6 +17,8 @@
     <CodeAnalysisRuleSet>..\common\GitHubVS.ruleset</CodeAnalysisRuleSet>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <CodeAnalysisIgnoreGeneratedCode>true</CodeAnalysisIgnoreGeneratedCode>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
```
Comment (shana) [inline src/GitHub.TeamFoundation.15/GitHub.TeamFoundation.15.csproj]: Same as above

File: src/GitHub.VisualStudio/GitHub.VisualStudio.csproj [hunk 10]
```diff
@@ -12,6 +12,8 @@
     <VsixType>v3</VsixType>
     <IsProductComponent>false</IsProductComponent>
     <ExtensionInstallationFolder>GitHub\GitHub</ExtensionInstallationFolder>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
```
Comment (shana) [inline src/GitHub.VisualStudio/GitHub.VisualStudio.csproj]: Same as above

File: src/GitHub.VisualStudio/app.config [hunk 11]
```diff
@@ -38,6 +38,14 @@
...
         <bindingRedirect oldVersion="0.0.0.0-4.0.1.0" newVersion="4.0.1.0" />
       </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="Microsoft.VisualStudio.ComponentModelHost" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-14.0.0.0" newVersion="14.0.0.0" />
+      </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="Microsoft.VisualStudio.Validation" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-14.0.0.0" newVersion="14.0.0.0" />
+      </dependentAssembly>
```
Comment (shana) [inline src/GitHub.VisualStudio/app.config]: Same as above

File: test/UnitTests/app.config [hunk 12]
```diff
@@ -34,6 +34,10 @@
         <assemblyIdentity name="Microsoft.VisualStudio.CoreUtility" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
         <bindingRedirect oldVersion="0.0.0.0-14.0.0.0" newVersion="14.0.0.0" />
       </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="Microsoft.VisualStudio.Validation" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-14.0.0.0" newVersion="14.0.0.0" />
+      </dependentAssembly>
```
Comment (shana) [inline test/UnitTests/app.config]: Same as above

**Comment by shana** [comment]
@jcansdale 

> NOTE: I've left in some of the dependentAssembly and NuGetPackageImportStamp junk that NuGet likes to add. Hopefully this will make the next time we update NuGet packages cleaner. We don't package the App.config files, so this shouldn't make any difference to the extension.

Yeah, existing things are totally ok. Not having the changes to junk in the commit for the fix is mainly for two things:

- The new things added and changed that made an hour backport operation into almost five hours, what with all the fighting the branch merges due to the junk lines changed being so close to the actual important changes.
- If someone does a blame on a csproj and sees a junk line changed but the commit message says "fix for libgit2", they can't tell whether that line was actually required for the fix or if it's just junk. If the commit message for stuff like this clearly says "clean up", then they know it's changes they don't need to look at, which helps a lot when backtracking through history.

**Comment by jcansdale** [comment]
@shana I've split the stuff that NuGet wanted to add into a separate "cosmetic" PR: #1325.

In the past I've tried to keep the diff nice and clean. In this case I wasn't sure what to do, because it feels like I've been fighting NuGet a lot recently. Hopefully this cosmetic PR will make future changes less of a pain.

**Pull request merged**